### DISCUSSION TITLE
(Why) does VIPS perform worse when called in parallel than when called sequentially?

### DISCUSSION THREAD

**Discussion by VivitionDeveloper** [libvips/libvips#4063] [Q&A] [answered]
## Questions, enhancements, tips, etc.

It seems that calling libvips multiple times in parallel performs worse than calling it multiple times sequentially. Why is this? Should I be wrapping my copy_memory/write functions in a mutex for better perfomance?

## Bug report

**Describe the bug**
When testing, libvips consistently performs worse when called in parallel over when called in sequence.

**To Reproduce**
1. Use any image.
2. Run the following C++ code:
```c++
		void BenchmarkVIPS() {
			printf("Testing VIPS parallel/seq \n");
			vips::VImage vipsImage = vips::VImage::new_from_file("inputbottomperunage.jpg");

			const auto fnApplyValueLUTUsingVips = [](const vips::VImage& vipsImage, const vips::VImage& mLUT) {
				// Clone the image.
				auto vipsImageCopy = vipsImage.copy_memory();

				// First convert to HSV.
				vipsImageCopy = vipsImageCopy.colourspace(VipsInterpretation::VIPS_INTERPRETATION_HSV);

				// Split the channels.
				auto vipsChannels = vipsImageCopy.bandsplit();

				// Verify the lookup table.
				assert(mLUT.bands() == 1);
				assert(mLUT.width() == 256);
				assert(mLUT.height() == 1);
				assert(mLUT.format() == VipsBandFormat::VIPS_FORMAT_UCHAR);

				// Apply the lookup table.
				vipsChannels[2] = vipsChannels[2].maplut(mLUT);

				// Merge the channels.
				vipsImageCopy = vipsChannels[0].bandjoin(vipsChannels[1]).bandjoin(vipsChannels[2]);

				// Convert back to BGR.
				vipsImageCopy = vipsImageCopy.colourspace(VipsInterpretation::VIPS_INTERPRETATION_sRGB);

				// Force flush by copying to memory.
				auto vipsImageCopyMemory = vipsImageCopy.copy_memory();


				return vipsImageCopyMemory;
			};


			const auto fnTest = [vipsImage](const auto& fnVIPS, const auto& lstVipsParams) {
				using ParamVips = typename std::decay_t<decltype(lstVipsParams)>::value_type;
				// Run the test 10 times. Plot a table with the results per iteration and the total time.
				size_t iTimesParVips = 0;
				size_t iTimesSeqVips = 0;
				printf("Times in MS\n");

				printf("%-10s | %-10s\n", "", "");
				printf("%10s | %10s\n", "Parallel", "Sequential");
				printf("%-10s | %-10s\n", "", "");
				for (size_t i = 0; i < 10; i++) {
					fw::Threading::ParallelRunner<ParamVips> prRunnerVips;
					for (const auto& paramVips : lstVipsParams) {
						prRunnerVips.addTask([vipsImage, fnVIPS](auto paramVips) {
							fnVIPS(vipsImage, paramVips);
						}, paramVips);
					}
					const auto tStartVips = std::chrono::high_resolution_clock::now();
					prRunnerVips.doRun();
					const auto tEndVips = std::chrono::high_resolution_clock::now();
					const size_t iTimeVips = std::chrono::duration_cast<std::chrono::milliseconds>(tEndVips - tStartVips).count();
					printf("%-10zu | ", iTimeVips);
					iTimesParVips += iTimeVips;

					const auto tStartSeqVips = std::chrono::high_resolution_clock::now();
					for (const auto& paramVips : lstVipsParams) {
						fnVIPS(vipsImage, paramVips);
					}
					const auto tEndSeqVips = std::chrono::high_resolution_clock::now();
					const size_t iTimeSeqVips = std::chrono::duration_cast<std::chrono::milliseconds>(tEndSeqVips - tStartSeqVips).count();
					printf("%-10zu\n", iTimeSeqVips);
					iTimesSeqVips += iTimeSeqVips;

				}
				printf("%-10s | %-10s\n", "", "");
				printf("%-10s | %-10s\n", "Total", "");
				printf("%-10zu | %-10zu\n", iTimesParVips, iTimesSeqVips);
			};

			size_t iCount = 100;
			std::list<double> lstAmount;
			for (size_t i = 0; i < iCount; i++) {
				lstAmount.push_back(1.0 + (1.0 / iCount * i));
			}

			const auto fnApplyUnsharpMaskUsingVips = [](const vips::VImage& vipsImage, const double dAmount) {
				auto vipsBlurred = vipsImage.gaussblur(9 / 2 + 1);// Smaller than opencv as opencv uses kernel, vips uses radius
				auto vipsSharpened = (dAmount + 1.0) * vipsImage - dAmount * vipsBlurred;
				auto vipsSharpenedClipped = (vipsSharpened < 0).ifthenelse(0, vipsSharpened);
				vipsSharpenedClipped = (vipsSharpenedClipped > 255).ifthenelse(255, vipsSharpenedClipped);
				return vipsSharpenedClipped.cast(VIPS_FORMAT_UCHAR).copy_memory();
			};

			printf("Testing unsharp mask\n");
			fnTest(fnApplyUnsharpMaskUsingVips, lstAmount);
		}
```
4. See the output printed to console

**Expected behavior**
I expected parallel to be faster, or at least similar in timing to sequential.

**Actual behavior**
Sequential consistently outperforms parallel. With this specific task it's not a lot, but with some other tasks (LUT) it's significant, sometimes over 30% difference. My output:
```
Testing VIPS parallel/seq
Testing unsharp mask
Times in MS
           |
  Parallel | Sequential
           |
47097      | 18308
13317      | 12855
14396      | 12272
13639      | 12441
15938      | 11757
15081      | 12536
15232      | 13348
15386      | 13828
13655      | 13076
13252      | 13414
           |
Total      |
176993     | 133835
```

**Screenshots**
None.

**Environment**
(please complete the following information)
 - OS: Windows 10/Windows server 2023
 - Vips: 8.15.1

**Additional context**
Not afaik.

**Comment by jcupitt** [comment]
Hello again, 

You'd need to post a complete example I could run, but it's probably that libvips automatically parallizes evaluation, so if you run more than one pipeline at once, you may overcommit on threads.

You can limit pipeline parallelism with `vips_concurrency_set()` or the `VIPS_CONCURRENCY` env var:

https://www.libvips.org/API/current/VipsOperation.html#vips-concurrency-set

Set it to 1 and libvips will run (mostly) single-threaded.

**Comment by VivitionDeveloper** [comment]
Thank you for your fast response.

The following code should be a complete example.
```cpp
#include <vips/vips8>
#include <chrono>
#include <thread>

int
main(int argc, char** argv) {
	VIPS_INIT(argv[0]);
	printf("Testing VIPS parallel/seq \n");
	vips::VImage vipsImage = vips::VImage::new_from_file("inputbottomperunage.jpg");

	const auto fnApplyValueLUTUsingVips = [](const vips::VImage& vipsImage, const vips::VImage& mLUT) {
		// Clone the image.
		auto vipsImageCopy = vipsImage.copy_memory();

		// First convert to HSV.
		vipsImageCopy = vipsImageCopy.colourspace(VipsInterpretation::VIPS_INTERPRETATION_HSV);

		// Split the channels.
		auto vipsChannels = vipsImageCopy.bandsplit();

		// Apply the lookup table.
		vipsChannels[2] = vipsChannels[2].maplut(mLUT);

		// Merge the channels.
		vipsImageCopy = vipsChannels[0].bandjoin(vipsChannels[1]).bandjoin(vipsChannels[2]);

		// Convert back to BGR.
		vipsImageCopy = vipsImageCopy.colourspace(VipsInterpretation::VIPS_INTERPRETATION_sRGB);

		// Force flush by copying to memory.
		auto vipsImageCopyMemory = vipsImageCopy.copy_memory();


		return vipsImageCopyMemory;
	};


	const auto fnTest = [vipsImage](const auto& fnVIPS, const auto& lstVipsParams) {
		using ParamVips = typename std::decay_t<decltype(lstVipsParams)>::value_type;
		// Run the test 10 times. Plot a table with the results per iteration and the total time.
		size_t iTimesParVips = 0;
		size_t iTimesSeqVips = 0;
		printf("Times in MS\n");

		printf("%-10s | %-10s\n", "", "");
		printf("%10s | %10s\n", "Parallel", "Sequential");
		printf("%-10s | %-10s\n", "", "");
		for (size_t i = 0; i < 10; i++) {
			std::list<std::thread> lstThreads;
			for (const auto& paramVips : lstVipsParams) {
				lstThreads.push_back(std::thread([&vipsImage, &fnVIPS, paramVips]() {
					fnVIPS(vipsImage, paramVips);
				}));
			}
			auto tStartVips = std::chrono::high_resolution_clock::now();
			for (auto& thread : lstThreads) {
				thread.join();
			}
			const auto tEndVips = std::chrono::high_resolution_clock::now();
			const size_t iTimeVips = std::chrono::duration_cast<std::chrono::milliseconds>(tEndVips - tStartVips).count();
			printf("%-10zu | ", iTimeVips);
			iTimesParVips += iTimeVips;

			const auto tStartSeqVips = std::chrono::high_resolution_clock::now();
			for (const auto& paramVips : lstVipsParams) {
				fnVIPS(vipsImage, paramVips);
			}
			const auto tEndSeqVips = std::chrono::high_resolution_clock::now();
			const size_t iTimeSeqVips = std::chrono::duration_cast<std::chrono::milliseconds>(tEndSeqVips - tStartSeqVips).count();
			printf("%-10zu\n", iTimeSeqVips);
			iTimesSeqVips += iTimeSeqVips;

		}
		printf("%-10s | %-10s\n", "", "");
		printf("%-10s | %-10s\n", "Total", "");
		printf("%-10zu | %-10zu\n", iTimesParVips, iTimesSeqVips);
	};

	size_t iCount = 100;
	std::list<double> lstAmount;
	for (size_t i = 0; i < iCount; i++) {
		lstAmount.push_back(1.0 + (1.0 / iCount * i));
	}

	const auto fnApplyUnsharpMaskUsingVips = [](const vips::VImage& vipsImage, const double dAmount) {
		auto vipsBlurred = vipsImage.gaussblur(9 / 2 + 1);// Smaller than opencv as opencv uses kernel, vips uses radius
		auto vipsSharpened = (dAmount + 1.0) * vipsImage - dAmount * vipsBlurred;
		auto vipsSharpenedClipped = (vipsSharpened < 0).ifthenelse(0, vipsSharpened);
		vipsSharpenedClipped = (vipsSharpenedClipped > 255).ifthenelse(255, vipsSharpenedClipped);
		return vipsSharpenedClipped.cast(VIPS_FORMAT_UCHAR).copy_memory();
	};

	printf("Testing unsharp mask\n");
	fnTest(fnApplyUnsharpMaskUsingVips, lstAmount);

	return 0;
}
```
This gives:
```
Testing VIPS parallel/seq
Testing unsharp mask
Times in MS
           |
  Parallel | Sequential
           |
9081       | 10509
12760      | 10740
11591      | 9939
12075      | 10148
12158      | 10313
11384      | 9704
13511      | 10815
13409      | 10652
12287      | 9517
12130      | 10424
           |
Total      |
120386     | 102761
```
I'll be honest, I don't fully understand what you mean by "overcommitting on threads", so I'm going to simply write out some thoughts. I take it that this means "you're asking your machine to do more than it can handle at once". Does this mean I should fabricate a work queue around libvips, to limit the amount of tasks I give to vips, in order for vips to work faster and retain a higher throughput? And as a thought that follows from this, If so, does that indicate an inefficiency in the VIPS threadpooling system, where a larger amount of items in the queue slows down throughput?

Or are threadpools per constructed VImage, and am I therefore running 100 threadpools at once? Testing with
```cpp
			vips_concurrency_set(1);
			std::list<std::thread> lstThreads;
			for (const auto& paramVips : lstVipsParams) {
				lstThreads.push_back(std::thread([&vipsImage, &fnVIPS, paramVips]() {
					fnVIPS(vipsImage, paramVips);
				}));
			}
			auto tStartVips = std::chrono::high_resolution_clock::now();
			for (auto& thread : lstThreads) {
				thread.join();
			}
			const auto tEndVips = std::chrono::high_resolution_clock::now();
			vips_concurrency_set(0);
```
for the parallel part gives:
```
Testing VIPS parallel/seq
Testing unsharp mask
Times in MS
           |
  Parallel | Sequential
           |
633        | 3383
467        | 3175
423        | 3218
397        | 3152
366        | 3213
422        | 3198
444        | 3253
420        | 3178
351        | 3109
342        | 3150
           |
Total      |
4265       | 32029
```


Eh, wow.... That's a whole lot faster.... What....

Even testing with:
```cpp
			vips_concurrency_set(1);
			std::list<std::thread> lstThreads;
			for (const auto& paramVips : lstVipsParams) {
				lstThreads.push_back(std::thread([&vipsImage, &fnVIPS, paramVips]() {
					fnVIPS(vipsImage, paramVips);
				}));
			}
			auto tStartVips = std::chrono::high_resolution_clock::now();
			for (auto& thread : lstThreads) {
				thread.join();
			}
			const auto tEndVips = std::chrono::high_resolution_clock::now();
			const size_t iTimeVips = std::chrono::duration_cast<std::chrono::milliseconds>(tEndVips - tStartVips).count();
			printf("%-10zu | ", iTimeVips);
			iTimesParVips += iTimeVips;

			const auto tStartSeqVips = std::chrono::high_resolution_clock::now();
			for (const auto& paramVips : lstVipsParams) {
				fnVIPS(vipsImage, paramVips);
			}
			const auto tEndSeqVips = std::chrono::high_resolution_clock::now();
			const size_t iTimeSeqVips = std::chrono::duration_cast<std::chrono::milliseconds>(tEndSeqVips - tStartSeqVips).count();
			printf("%-10zu\n", iTimeSeqVips);
			iTimesSeqVips += iTimeSeqVips;
			vips_concurrency_set(0);
```
(so sequentially also single-threaded)
gives me:
```
esting VIPS parallel/seq
Testing unsharp mask
Times in MS
           |
  Parallel | Sequential
           |
601        | 3313
373        | 3265
436        | 3165
404        | 3186
382        | 3202
400        | 3196
395        | 3256
407        | 3187
376        | 3225
343        | 3114
           |
Total      |
4117       | 32109
```
So, when executing sequentially it doesn't matter if I give it a concurrency of 1, and when executing 100 operations in parallel setting a concurrency of 1 is way faster? And setting a concurrency of 1 does not affect sequential?

I'm amazed at this result. I'm running this in a hyper-v environment, the CPU is an AMD EPYC 7443P 24-core (48 virtual cores), with 293GB of RAM, according to windows. The VHDX from which the image is loaded is on a Samsung PM1733 2,5" 7,68TB.

From this data I'd assume I should set VIPS_CONCURRENCY to 1 for better performance? I can't quite imagine that to be right...

**Reply by VivitionDeveloper** [reply]
Oh, wait. There seems to be a bug in vips_concurrency_set_0. Running
```cpp
	printf("Concurrency: %d\n", vips_concurrency_get());
	vips_concurrency_set(0);
	printf("Concurrency: %d\n", vips_concurrency_get());
	vips_concurrency_set(1);
	printf("Concurrency: %d\n", vips_concurrency_get());
	vips_concurrency_set(0);
	printf("Concurrency: %d\n", vips_concurrency_get());
```
gives me:
```
Concurrency: 48
Concurrency: 48
Concurrency: 1
Concurrency: 1
```
where all but the last one is expected. I expected the last line to give 48 again. Am I misreading the documentation?

**Reply by VivitionDeveloper** [reply]
And testing multiple concurrency values using the following code:
```cpp
for (size_t i = 0; i < 10; i++) {
			vips_concurrency_set(i * 4);
			printf("Concurrency: %d\n", vips_concurrency_get());
			std::list<std::thread> lstThreads;
			for (const auto& paramVips : lstVipsParams) {
				lstThreads.push_back(std::thread([&vipsImage, &fnVIPS, paramVips]() {
					fnVIPS(vipsImage, paramVips);
				}));
			}
			auto tStartVips = std::chrono::high_resolution_clock::now();
			for (auto& thread : lstThreads) {
				thread.join();
			}
			const auto tEndVips = std::chrono::high_resolution_clock::now();
			const size_t iTimeVips = std::chrono::duration_cast<std::chrono::milliseconds>(tEndVips - tStartVips).count();
			printf("%-10zu | ", iTimeVips);
			iTimesParVips += iTimeVips;

			const auto tStartSeqVips = std::chrono::high_resolution_clock::now();
			for (const auto& paramVips : lstVipsParams) {
				fnVIPS(vipsImage, paramVips);
			}
			const auto tEndSeqVips = std::chrono::high_resolution_clock::now();
			const size_t iTimeSeqVips = std::chrono::duration_cast<std::chrono::milliseconds>(tEndSeqVips - tStartSeqVips).count();
			printf("%-10zu\n", iTimeSeqVips);
			iTimesSeqVips += iTimeSeqVips;
		}
```
indicates to me that in both cases, a concurrency of 1 is indeed best:
```
Testing VIPS parallel/seq
Concurrency: 48
Concurrency: 48
Concurrency: 1
Concurrency: 1
Testing unsharp mask
Times in MS
           |
  Parallel | Sequential
           |
Concurrency: 1
641        | 3305 <- first run is always a bit slower than later runs. Cache warming or so?
Concurrency: 4
847        | 1258
Concurrency: 8
1246       | 1210
Concurrency: 12
2312       | 2692
Concurrency: 16
3066       | 3026
Concurrency: 20
3851       | 3695
Concurrency: 24
5230       | 4560
Concurrency: 28
8018       | 5954
Concurrency: 32
10699      | 7628
Concurrency: 36
10808      | 8664
           |
Total      |
46718      | 41992
```

**Reply by VivitionDeveloper** [reply]
I'm seeing roughly the same on a different system with 8 cores / 16 threads:
```
Testing VIPS parallel/seq
Testing unsharp mask
Times in MS
           |
  Parallel | Sequential
           |
Concurrency: 16
2094       | 398
Concurrency: 1
209        | 541
Concurrency: 2
222        | 301
Concurrency: 3
224        | 227
Concurrency: 4
241        | 194
Concurrency: 5
230        | 166
Concurrency: 6
338        | 170
Concurrency: 7
294        | 148
Concurrency: 8
284        | 154
Concurrency: 9
356        | 170
Concurrency: 10
302        | 160
Concurrency: 11
413        | 185
Concurrency: 12
391        | 190
Concurrency: 13
470        | 196
Concurrency: 14
470        | 270
Concurrency: 15
480        | 246
Concurrency: 16
514        | 248
           |
Total      |
7532       | 3964
```

I'm now running 12 tasks in parallel/sequentially. This is on a ryzen 7 2700, not virtualized. It seems that in sequential operation, 7 is fastst. Is it safe to say that VIPS has the highest throughput when the total number of tasks running concurrently is equal to the number of physical cores, minus one for windows/IO? If so, that would indicate that placing a static mutex around all my processing calls (write, copy_memory, etc) and reducing VIPS_CONCURRENCY to the number of physical cores, would increase my performance by (248-148)/148=67%?

**Comment by jcupitt** [comment]
It depends (of course!). Random thoughts:

Windows is pretty slow with threads, you'll get a different result on macos or linux.

The libvips concurrency setting is the maximum size of a threadpool, not the actual size. Threadpools size up and down automatically with load and aim to minimise stalling.

Threadpools are created (and recycled) per evaluation pipeline, so lot of images won't cause overcommitting, but lots of pipelines in parallel can.

You'll see very different results with different operation pipelines, of course. A usharp operation with a large sigma is a good test, but it won't hit your FP units, so you won't see any benefit from hyperthreading.

"Overcommitting" here means having many more active threads than available CPU cores. The OS will timeshare cores over threads in some way (round-robin? something like that), and that will continually empty the caches as cores hop between tasks. It's best to have number-of-threads-with-useful-work-to-do == number-of-cpu-cores, with the exception that a mix of int and fp work can make a degree of hyperthreading worthwhile.

libvips runs several extra threads beyond the number set by `VIPS_CONCURRENCY`, so it's not an exact guide.

For speed, it's also worthwhile to think about throughput versus latency. Do you want a single result as quickly as possible? Or do you want the most results per minute, and don't care about how long each one takes?

**Reply by jcupitt** [reply, marked as answer]
... and if you're aiming for throughput on a large set of images, single-thread mode and `$cores` pipelines running at once will always be the best solution, whatever image processing system you are using. libvips works well in one-thread mode since it has relatively low memory use. You won't run out of RAM almost no matter how many pipelines you run at once or how large your images are.

tldr: if you want to process a single image quickly, then set `VIPS_CONCURRENCY` to 0 and run a one pipeline. If you want to batch process a large set of images, set it to 1 and run a pool of workers sized to just below the number of physical cores.

**Reply by VivitionDeveloper** [reply]
Awesome, thanks! We're going to try this out and see how much we actually gain in our production environment. We're mostly talking about batch processing, so setting it to 1 should give quite a performance boost.
